File: bastion-mcp-server/src/config/merge.ts
/**
 * Strategic merge for layered YAML/JSON configuration.
 *
 * Follows Kubernetes strategic merge patch semantics:
 *   - Scalars: overlay replaces base
 *   - Objects: recursive merge
 *   - Arrays of objects with a `name` field: merge by name
 *   - Scalar arrays: overlay replaces base entirely
 *   - null/undefined overlay values: base preserved
 */

/**
 * Deep-merge `overlay` onto `base`, returning a new object.
 * Neither input is mutated.
 */
export function strategicMerge(base: unknown, overlay: unknown): unknown {
  if (overlay === null || overlay === undefined) {
    return base;
  }

  if (base === null || base === undefined) {
    return overlay;
  }

  // Both are arrays
  if (Array.isArray(base) && Array.isArray(overlay)) {
    return mergeArrays(base, overlay);
  }

  // Both are plain objects
  if (isPlainObject(base) && isPlainObject(overlay)) {
    return mergeObjects(
      base as Record<string, unknown>,
      overlay as Record<string, unknown>
    );
  }

  // Scalar: overlay wins
  return overlay;
}

function mergeObjects(
  base: Record<string, unknown>,
  overlay: Record<string, unknown>
): Record<string, unknown> {
  const result: Record<string, unknown> = { ...base };

  for (const key of Object.keys(overlay)) {
    const overlayVal = overlay[key];
    if (overlayVal === null || overlayVal === undefined) {
      // null/undefined in overlay → keep base value
      continue;
    }
    if (key in base) {
      result[key] = strategicMerge(base[key], overlayVal);
    } else {
      result[key] = overlayVal;
    }
  }

  return result;
}

function mergeArrays(base: unknown[], overlay: unknown[]): unknown[] {
  // Check if both arrays contain objects with a `name` field (named resources)
  if (isNamedObjectArray(base) && isNamedObjectArray(overlay)) {
    return mergeNamedArrays(
      base as Array<Record<string, unknown>>,
      overlay as Array<Record<string, unknown>>
    );
  }

  // Scalar arrays: overlay replaces entirely
  return overlay;
}

function mergeNamedArrays(
  base: Array<Record<string, unknown>>,
  overlay: Array<Record<string, unknown>>
): Array<Record<string, unknown>> {
  const result = base.map((item) => ({ ...item }));

  for (const overlayItem of overlay) {
    const name = overlayItem.name;
    const existingIndex = result.findIndex((item) => item.name === name);

    if (existingIndex >= 0) {
      result[existingIndex] = strategicMerge(result[existingIndex], overlayItem) as Record<string, unknown>;
    } else {
      result.push(overlayItem);
    }
  }

  return result;
}

function isPlainObject(value: unknown): boolean {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function isNamedObjectArray(arr: unknown[]): boolean {
  if (arr.length === 0) return false;
  return arr.every(
    (item) => isPlainObject(item) && 'name' in (item as Record<string, unknown>)
  );
}

File: internal/config/merge.go
package config

// StrategicMerge performs a Kubernetes-style strategic merge of overlay onto base.
//   - Scalars: overlay replaces base
//   - Maps: recursive merge
//   - Slices of maps with a "name" key: merge by name
//   - Scalar slices: overlay replaces entirely
//   - nil overlay: base preserved
func StrategicMerge(base, overlay interface{}) interface{} {
	if overlay == nil {
		return base
	}
	if base == nil {
		return overlay
	}

	baseMap, baseIsMap := toStringMap(base)
	overlayMap, overlayIsMap := toStringMap(overlay)
	if baseIsMap && overlayIsMap {
		return mergeMaps(baseMap, overlayMap)
	}

	baseSlice, baseIsSlice := toSlice(base)
	overlaySlice, overlayIsSlice := toSlice(overlay)
	if baseIsSlice && overlayIsSlice {
		return mergeSlices(baseSlice, overlaySlice)
	}

	return overlay
}

func mergeMaps(base, overlay map[string]interface{}) map[string]interface{} {
	result := make(map[string]interface{}, len(base))
	for k, v := range base {
		result[k] = v
	}
	for k, v := range overlay {
		if v == nil {
			continue
		}
		if existing, ok := result[k]; ok {
			result[k] = StrategicMerge(existing, v)
		} else {
			result[k] = v
		}
	}
	return result
}

func mergeSlices(base, overlay []interface{}) interface{} {
	if isNamedSlice(base) && isNamedSlice(overlay) {
		return mergeNamedSlices(base, overlay)
	}
	return overlay
}

func mergeNamedSlices(base, overlay []interface{}) []interface{} {
	result := make([]interface{}, len(base))
	for i, v := range base {
		if m, ok := toStringMap(v); ok {
			cp := make(map[string]interface{}, len(m))
			for k, val := range m {
				cp[k] = val
			}
			result[i] = cp
		} else {
			result[i] = v
		}
	}

	for _, ov := range overlay {
		om, _ := toStringMap(ov)
		name := om["name"]
		found := false
		for i, bv := range result {
			bm, _ := toStringMap(bv)
			if bm["name"] == name {
				result[i] = StrategicMerge(bv, ov)
				found = true
				break
			}
		}
		if !found {
			result = append(result, ov)
		}
	}
	return result
}

func isNamedSlice(s []interface{}) bool {
	if len(s) == 0 {
		return false
	}
	for _, item := range s {
		m, ok := toStringMap(item)
		if !ok {
			return false
		}
		if _, hasName := m["name"]; !hasName {
			return false
		}
	}
	return true
}

func toStringMap(v interface{}) (map[string]interface{}, bool) {
	m, ok := v.(map[string]interface{})
	return m, ok
}

func toSlice(v interface{}) ([]interface{}, bool) {
	s, ok := v.([]interface{})
	return s, ok
}
